src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -1458,6 +1457,39 @@ internal virtual bool TryApplyChanges(Solution newSolution, IProgress<CodeAnalys
             }
         }
 
+        private void ApplyDocumentsInfoChange(ImmutableArray<ProjectChanges> projectChanges)
+        {
+            using var _1 = PooledHashSet<DocumentId>.GetInstance(out var infoChangedDocumentIds);
+            using var _2 = PooledHashSet<Document>.GetInstance(out var infoChangedNewDocument);
+            foreach (var projectChange in projectChanges)
+            {
+                foreach (var docId in projectChange.GetChangedDocuments())
+                {
+                    var oldSolution = projectChange.OldProject.Solution;
+                    var oldDoc = projectChange.OldProject.GetRequiredDocument(docId);
+                    var newDoc = projectChange.NewProject.GetRequiredDocument(docId);
+                    if (oldDoc.HasInfoChanged(newDoc))
+                    {
+                        var linkedDocumentIds = oldSolution.GetRequiredDocument(docId).GetLinkedDocumentIds();
```

Comment:
Since you've already got this list, consider just adding all of them in line 1478. That then removes the need for the .All() check and you can just check the Id on line 1471 before doing anything further. The idea being rather than each time we process re-checking if this is a known duplicate, just add to the hash set the document Ids we've already "covered". Might consume a hair more memory but would reduce overall work.